### PR TITLE
[ticket/15766] Check for role on changing permission settings

### PR DESCRIPTION
**Pull request by mrgoldy**
Checklist:

- [x] Correct branch: master for new features; 3.3.x & 3.2.x for fixes
- [x] Tests pass
- [x] Code follows coding guidelines: [master](https://area51.phpbb.com/docs/dev/master/development/coding_guidelines.html), [3.3.x](https://area51.phpbb.com/docs/dev/3.3.x/development/coding_guidelines.html) and [3.2.x](https://area51.phpbb.com/docs/dev/3.2.x/development/coding_guidelines.html)
- [x] Commit follows commit message [format](https://area51.phpbb.com/docs/dev/3.3.x/development/git.html)

Tracker ticket (set the ticket ID to **your ticket ID**):

https://tracker.phpbb.com/browse/PHPBB3-15766

Perhaps I should explain a bit better what this actually does. 
When you change a permission, it checks if the new group of permissions matches a permission role and indicates it as such in the dropdown. Rather than always defaulting to ‘No role assigned...’

### CODE REVIEW

File: phpBB/adm/style/permissions.js [hunk 1]
```diff
@@ -278,11 +278,26 @@ function reset_role(id) {
...
 
-	t.options[0].selected = true;
+	var parent = t.parentNode,
+		roleId = match_role_settings(id.replace('role', 'perm')),
+		text = no_role_assigned,
+		index = 0;
+
+	if (roleId) {
+		for (var i = 0; i < t.options.length; i++) {
+			if (t.options[i].value == roleId) {
```
Comment (marc1706) [inline phpBB/adm/style/permissions.js]:
```suggestion
			if (t.options[i].value === roleId) {
```

File: phpBB/adm/style/permissions.js [hunk 2]
```diff
@@ -302,3 +317,36 @@ function set_role_settings(role_id, target_id) {
 		mark_one_option(target_id, r, (settings[r] === 1) ? 'y' : 'n');
 	}
 }
+
+function match_role_settings(id)
```
Comment (marc1706) [inline phpBB/adm/style/permissions.js]: Please add docblocks for newly added functions.

File: phpBB/adm/style/permissions.js [hunk 3]
```diff
@@ -302,3 +317,36 @@ function set_role_settings(role_id, target_id) {
...
+{
+	var fs = document.getElementById(id),
```
Comment (marc1706) [inline phpBB/adm/style/permissions.js]: I understand that these might be in line with some of the older JS but you should use variable names that actually mean something. These are not helpful at all and remind me a bit of really old C code. ;)

File: phpBB/adm/style/permissions.js [hunk 4]
```diff
@@ -302,3 +317,36 @@ function set_role_settings(role_id, target_id) {
...
+		cbs = fs.getElementsByTagName('input'),
+		set = {};
+
+	for (var i = 0; i < cbs.length; i++) {
+		var matches = cbs[i].id.match(/setting\[\d+]\[\d+]\[([a-z_]+)]/);
+
+		if (matches !== null && cbs[i].checked && cbs[i].value !== '-1') {
+			set[matches[1]] = parseInt(cbs[i].value);
```
Comment (marc1706) [inline phpBB/adm/style/permissions.js]: Should add the radix parameter:
```suggestion
			set[matches[1]] = parseInt(cbs[i].value, 10);
```

File: phpBB/adm/style/permissions.js [hunk 5]
```diff
@@ -302,3 +317,36 @@ function set_role_settings(role_id, target_id) {
...
+	{
+		if (sort_and_stringify(role_options[r]) === set) {
+			return r;
+		}
+	}
+
+	return 0;
+}
+
+function sort_and_stringify(obj) {
```
Comment (marc1706) [inline phpBB/adm/style/permissions.js]: Same as above.

**Comment by 3D-I** [comment]
Wait.. I haven't tried this one but a question arose: 

* Is the ‘No role assigned...’ thinghy not available any more here?

**Comment by marc1706** [comment]
It is still available, it'll however display the role you picked when selecting the same settings as a current role.

**Pull request merged**